File: docs/nifi-analyzer/src/parsers/nel/parser.js
// ================================================================
// nel/parser.js — Recursive-descent NEL expression parser
// Extracted from monolith lines 3124-3175
// ================================================================

import { tokenizeNELChain } from './tokenizer.js';
import { resolveNELVariableContext } from './variable-context.js';
import { resolveNELArg } from './arg-resolver.js';
import { applyNELFunction } from './function-handlers.js';

/**
 * Parse a single NEL expression (the content inside ${...}) into
 * PySpark column operations or Python string operations.
 *
 * Handles:
 *   - System functions: now(), UUID(), hostname(), nextInt(), random(), literal()
 *   - Math namespace: math:abs(), math:ceil(), math:floor(), math:round()
 *   - Variable references with context classification (secret/config/column)
 *   - Chained function calls via colon-delimited syntax
 *
 * @param {string} expr - Expression content (without surrounding ${})
 * @param {string} mode - 'col' for PySpark column ops, 'python' for string ops
 * @returns {string} Translated PySpark or Python code
 */
export function parseNELExpression(expr, mode) {
  // Tokenize: split on colon-delimited function calls, respecting nested parens
  var parts = tokenizeNELChain(expr);
  if (parts.length === 0) return '""';

  var base = parts[0].trim();
  var chain = parts.slice(1);

  // Resolve the base reference
  var result;
  // System functions (no base variable)
  if (/^now\(\)$/i.test(base)) {
    result = mode === 'col' ? 'current_timestamp()' : 'datetime.now()';
  } else if (/^UUID\(\)$/i.test(base)) {
    result = mode === 'col' ? 'expr("uuid()")' : 'str(uuid.uuid4())';
  } else if (/^hostname\(\)$/i.test(base)) {
    result = mode === 'col' ? 'lit(spark.conf.get("spark.databricks.clusterUsageTags.clusterName", "unknown"))' : 'socket.gethostname()';
  } else if (/^nextInt\(\)$/i.test(base) || /^random\(\)$/i.test(base)) {
    result = mode === 'col' ? '(rand() * 2147483647).cast("int")' : 'random.randint(0, 2147483647)';
  } else if (/^literal\('((?:[^'\\]|\\.)*)'\)$/i.test(base)) {
    var litVal = base.match(/^literal\('((?:[^'\\]|\\.)*)'\)$/i)[1];
    var escaped = litVal.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n');
    result = mode === 'col' ? 'lit("' + escaped + '")' : '"' + escaped + '"';
  } else if (/^literal\((\d+)\)$/i.test(base)) {
    var litNum = base.match(/^literal\((\d+)\)$/i)[1];
    result = mode === 'col' ? 'lit(' + litNum + ')' : litNum;
  // NOTE: math:* regexes use /^math:func\((.+)\)$/ — the $ anchor forces the
  // literal \) to match the LAST ')' in the string, so .+ (greedy) correctly
  // captures everything between the first '(' and last ')'. This handles
  // single-level nesting like math:ceil(abs(-5)) -> captures "abs(-5)".
  } else if (/^math:abs\((.+)\)$/i.test(base)) {
    var absArg = base.match(/^math:abs\((.+)\)$/i)[1];
    result = mode === 'col' ? 'abs(' + resolveNELArg(absArg, mode) + ')' : 'abs(' + resolveNELArg(absArg, mode) + ')';
  } else if (/^math:ceil\((.+)\)$/i.test(base)) {
    result = mode === 'col' ? 'ceil(' + resolveNELArg(base.match(/^math:ceil\((.+)\)$/i)[1], mode) + ')' : 'math.ceil(' + resolveNELArg(base.match(/^math:ceil\((.+)\)$/i)[1], mode) + ')';
  } else if (/^math:floor\((.+)\)$/i.test(base)) {
    result = mode === 'col' ? 'floor(' + resolveNELArg(base.match(/^math:floor\((.+)\)$/i)[1], mode) + ')' : 'math.floor(' + resolveNELArg(base.match(/^math:floor\((.+)\)$/i)[1], mode) + ')';
  } else if (/^math:round\((.+)\)$/i.test(base)) {
    result = mode === 'col' ? 'round(' + resolveNELArg(base.match(/^math:round\((.+)\)$/i)[1], mode) + ')' : 'round(' + resolveNELArg(base.match(/^math:round\((.+)\)$/i)[1], mode) + ')';
  } else {
    // Variable reference
    var ctx = resolveNELVariableContext(base);
    if (mode === 'col') {
      result = ctx.type === 'column' ? 'col("' + base + '")' : ctx.code;
    } else {
      result = ctx.type === 'column' ? '_attrs.get("' + base + '", "")' :
               ctx.type === 'secret' ? ctx.code : ctx.code;
    }
  }

  // Apply chained function calls
  for (var i = 0; i < chain.length; i++) {
    result = applyNELFunction(result, chain[i].trim(), mode);
  }

  return result;
}

File: docs/nifi-analyzer/backend/app/engines/parsers/nel/parser.py
"""NEL recursive-descent parser — ported from nel/parser.js.

Parses NiFi Expression Language expressions into PySpark column ops or Python string ops.
"""

from __future__ import annotations

import re

from app.engines.parsers.nel.functions import (
    apply_nel_function,
    resolve_nel_arg,
    resolve_variable_context,
)
from app.engines.parsers.nel.tokenizer import tokenize_nel_chain


def parse_nel_expression(expr: str, mode: str = "col") -> str:
    """Parse a single NEL expression (content inside ${...}) into PySpark or Python code.

    Args:
        expr: Expression content without surrounding ${}.
        mode: 'col' for PySpark column ops, 'python' for string ops.

    Returns:
        Translated PySpark or Python code string.
    """
    parts = tokenize_nel_chain(expr)
    if not parts:
        return '""'

    base = parts[0].strip()
    chain = parts[1:]

    # System functions (no base variable)
    if re.match(r"^now\(\)$", base, re.IGNORECASE):
        result = "current_timestamp()" if mode == "col" else "datetime.now()"
    elif re.match(r"^UUID\(\)$", base, re.IGNORECASE):
        result = 'expr("uuid()")' if mode == "col" else "str(uuid.uuid4())"
    elif re.match(r"^hostname\(\)$", base, re.IGNORECASE):
        result = (
            'lit(spark.conf.get("spark.databricks.clusterUsageTags.clusterName", "unknown"))'
            if mode == "col"
            else "socket.gethostname()"
        )
    elif re.match(r"^nextInt\(\)$|^random\(\)$", base, re.IGNORECASE):
        result = '(rand() * 2147483647).cast("int")' if mode == "col" else "random.randint(0, 2147483647)"
    elif m := re.match(r"^literal\('((?:[^'\\]|\\.)*)'\)$", base, re.IGNORECASE):
        lit_val = m.group(1).replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
        result = f'lit("{lit_val}")' if mode == "col" else f'"{lit_val}"'
    elif m := re.match(r"^literal\((\d+)\)$", base, re.IGNORECASE):
        result = f"lit({m.group(1)})" if mode == "col" else m.group(1)
    # Math namespace
    elif m := re.match(r"^math:abs\((.+)\)$", base, re.IGNORECASE):
        arg = resolve_nel_arg(m.group(1), mode)
        result = f"abs({arg})"
    elif m := re.match(r"^math:ceil\((.+)\)$", base, re.IGNORECASE):
        arg = resolve_nel_arg(m.group(1), mode)
        result = f"ceil({arg})" if mode == "col" else f"math.ceil({arg})"
    elif m := re.match(r"^math:floor\((.+)\)$", base, re.IGNORECASE):
        arg = resolve_nel_arg(m.group(1), mode)
        result = f"floor({arg})" if mode == "col" else f"math.floor({arg})"
    elif m := re.match(r"^math:round\((.+)\)$", base, re.IGNORECASE):
        arg = resolve_nel_arg(m.group(1), mode)
        result = f"round({arg})"
    else:
        # Variable reference
        ctx = resolve_variable_context(base)
        if mode == "col":
            result = f'col("{base}")' if ctx["type"] == "column" else ctx["code"]
        else:
            if ctx["type"] == "column":
                result = f'_attrs.get("{base}", "")'
            else:
                result = ctx["code"]

    # Apply chained function calls
    for segment in chain:
        result = apply_nel_function(result, segment.strip(), mode)

    return result


def translate_nel_string(text: str, mode: str = "col") -> str:
    """Translate all ${...} expressions in a string to PySpark/Python code.

    Non-expression text is preserved as string literals. Multiple expressions
    are concatenated.
    """
    parts: list[str] = []
    pos = 0

    while pos < len(text):
        idx = text.find("${", pos)
        if idx < 0:
            # Remaining text is literal
            remaining = text[pos:]
            if remaining:
                escaped = remaining.replace("\\", "\\\\").replace('"', '\\"')
                parts.append(f'lit("{escaped}")' if mode == "col" else f'"{escaped}"')
            break

        # Literal text before ${
        if idx > pos:
            literal = text[pos:idx]
            escaped = literal.replace("\\", "\\\\").replace('"', '\\"')
            parts.append(f'lit("{escaped}")' if mode == "col" else f'"{escaped}"')

        # Find matching }
        depth = 0
        end = idx + 2
        while end < len(text):
            if text[end] == "{":
                depth += 1
            elif text[end] == "}":
                if depth == 0:
                    break
                depth -= 1
            end += 1

        if end >= len(text):
            # Unmatched ${, treat rest as literal
            remaining = text[idx:]
            escaped = remaining.replace("\\", "\\\\").replace('"', '\\"')
            parts.append(f'lit("{escaped}")' if mode == "col" else f'"{escaped}"')
            break

        expr = text[idx + 2 : end]
        parts.append(parse_nel_expression(expr, mode))
        pos = end + 1

    if not parts:
        return '""'
    if len(parts) == 1:
        return parts[0]
    if mode == "col":
        return "concat(" + ", ".join(parts) + ")"
    return " + ".join(parts)
